### ISSUE TITLE
GetViewport().GetCamera3D() returns instance of the script attached to camera in C#

### ISSUE THREAD

**Issue by markotaht** [godotengine/godot#79538]
### Godot version

4.1

### System information

Windows 10

### Issue description

I have a control script on Camera that allows me to move the camera. In another script i need to access the camera to do raycasting from mouse position. GetViewport().GetCamera3D() Tries ot return the instance of the moving scrip attached to the camera instead of returnign the Camera3D object that i need.
```

E 0:00:00:0872   :0 @ System.Object System.Runtime.CompilerServices.CastHelpers.ChkCastAny(System.Void* , System.Object ): System.InvalidCastException: Unable to cast object of type 'CameraController' to type 'Godot.Camera3D'.
  <C++ Error>    System.InvalidCastException
  <C++ Source>   :0 @ System.Object System.Runtime.CompilerServices.CastHelpers.ChkCastAny(System.Void* , System.Object )
  <Stack Trace>  :0 @ System.Object System.Runtime.CompilerServices.CastHelpers.ChkCastAny(System.Void* , System.Object )
                 VariantUtils.generic.cs:385 @ T Godot.NativeInterop.VariantUtils.ConvertTo<T >(Godot.NativeInterop.godot_variant& )
                 map_controller_ScriptProperties.generated.cs:14 @ Boolean map_controller.SetGodotClassPropertyValue(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:57 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Set(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant* )
```

If i try changeing the variable type into CameraController then i get tha reverse error saying that im trying to convert Godot.Camera3D into CameraController.

Seems like the API for C# is a bit broken.
 

### Steps to reproduce

Step 1: Add Camera3D to the scene.
Step 2: Create some script that controls camera movement, make it extend Node3D.
Step 3: Create another Node with Script to the scene.
Step 4: Try to save result of GetViewPort().GetCamera3D() into a variablt Typed Camera3D.

```
public partial class CameraController : Node3D
{
	// Declare member variables here. Examples:
	// private int a = 2;
	// private string b = "text";

	[Export]
	private float camera_speed = 4;
	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
	}

  // Called every frame. 'delta' is the elapsed time since the previous frame.
	public override void _Process(double delta)
	{      
		Viewport viewport = GetViewport();
		Vector2 size = viewport.GetVisibleRect().Size;
		Vector2 mouse = viewport.GetMousePosition();

		Vector3 offset = Vector3.Zero;
		if(mouse.X < 25){
			offset.X = -1;
		}else if (mouse.X > size.X - 25){
			offset.X = 1;
		}

		if(mouse.Y < 25){
			offset.Y = 1;
		}else if (mouse.Y > size.Y - 25){
			offset.Y = -1;
		}

		Translate(offset * Convert.ToSingle(delta) * camera_speed);
	}
}
```

```
public partial class map_controller : Node3D
{
	[Export]
	private Camera3D camera {get; set;}

	// Declare member variables here. Examples:
	// private int a = 2;
	// private string b = "text";

	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
		camera = GetViewport().GetCamera3D(); //This line is causing the error
	}

	//  // Called every frame. 'delta' is the elapsed time since the previous frame.
  //  public override void _Process(double delta)
  //  {
  //  }

}
```

### Minimal reproduction project

"N/A"

**Comment by AThousandShips** [comment]
You probably need to have your script extend `Camera3D`, is there any specific reason you are extending `Node3D` instead?

**Comment by markotaht** [comment]
Yep that worked. It just seems strange to me. Like if im asking for Camera3D i should be getting the Camera3D component from the object. Could be my Unity background. Need to unlearn many things.

**Comment by AThousandShips** [comment]
It returns the instance, the script isn't the instance but the instance is the type defined by the script

**Comment by markotaht** [comment]
Still weird to me. But hey thats the years of unity development. There its simple. You add as many scripts to a object you want and when you ask you get only that script instance not the whole thing. But you can access the entire object through the script by asking for the object.

**Comment by AThousandShips** [comment]
Yep, need to unlearn that and learn the Godot structure

**Comment by markotaht** [comment]
I have been trying to find some good documentation but most i can find are form 2016 and not up anymore. Any good source to read up on godot and its structures ?

**Comment by AThousandShips** [comment]
The [docs](https://docs.godotengine.org/)

**Comment by markotaht** [comment]
But where can i find something that explains how this component structure works? Does adding script make the entire object instance of that script? Or is the script just a sub object? Like what is happenign behind the scene? In unity its simple. Each script is a MonoBehaviour that is added to a list that needs to be updated every cycle. And objects themselves are just containers for those scripts.

**Comment by AThousandShips** [comment]
Probably under [scripting](https://docs.godotengine.org/en/stable/tutorials/scripting/index.html) please take a look around

**Comment by RedworkDE** [comment]
> component structure works

Godot doesn't use an ECS like Unity, so you won't find information on how it works anywhere (except in the godot-ecs repo, which is not an official thing tho)

**Comment by raulsntos** [comment]
I don't think they meant ECS, I think they meant _composition_. It seems they thought attaching scripts worked like adding "child-like components" to a Node, so basically _composition_ over _inheritance_.

I think Godot works more similar to Unreal Engine in this regard, where the script class **is** the node class. I can see how this may be confusing coming from Unity since both engines use the word _script_, and the word _attach_ probably doesn't help.

> Does adding script make the entire object instance of that script? Or is the script just a sub object? Like what is happenign behind the scene?

Yes, @AThousandShips already clarified this, but to reiterate you can think of adding scripts as changing the type of the object. When a node is instantiated it uses the C# class of the script to do so. So you could say that adding the script makes the object be an instance of that script.

It's not the first time users have been confused by this though (see https://github.com/godotengine/godot/issues/79080), which makes me think we need to better document this.
